File: mobile_apps/job_finder_standalone_latest/app/src/main/java/com/vc/jobfinder/domain/CompanyMatcher.kt
package com.vc.jobfinder.domain

import javax.inject.Inject
import javax.inject.Singleton

/**
 * Normalizes raw sponsor names ("Google LLC") and matches against canonical
 * company names from companies.yaml. Returns null if no confident match.
 */
@Singleton
class CompanyMatcher @Inject constructor(
    private val companiesProvider: CompaniesProvider,
) {
    private val suffixes = Regex(
        "\\b(inc|incorporated|ltd|limited|llc|llp|pvt|private|corp|corporation|" +
        "company|co|gmbh|sa|ag|plc|labs|technologies|tech|systems|solutions|" +
        "software|india|global)\\b\\.?",
        RegexOption.IGNORE_CASE
    )
    private val nonAlnum = Regex("[^a-z0-9]+")
    private val whitespace = Regex("\\s+")

    private fun normalize(name: String?): String {
        if (name.isNullOrBlank()) return ""
        var s = name.lowercase()
        s = suffixes.replace(s, "")
        s = nonAlnum.replace(s, " ").trim()
        s = whitespace.replace(s, " ")
        return s
    }

    private val index: Map<String, String> by lazy {
        companiesProvider.companies.associate { normalize(it.name) to it.name }
    }

    fun match(raw: String?): String? {
        val norm = normalize(raw)
        if (norm.isEmpty()) return null

        index[norm]?.let { return it }

        val candidates = mutableListOf<Pair<Int, String>>()
        for ((trackedNorm, canonical) in index) {
            if (trackedNorm.isEmpty()) continue
            if (trackedNorm in norm || norm in trackedNorm) {
                val score = minOf(trackedNorm.length, norm.length)
                candidates.add(score to canonical)
            }
        }
        return candidates.maxByOrNull { it.first }?.second
    }
}

interface CompaniesProvider {
    val companies: List<Company>
}

File: mobile_apps/agentic_job_finder_mobile_latest/src/competitions/_company_matcher.py
"""Match raw sponsor names to canonical companies in companies.yaml.

Normalizes by lowercasing and stripping common legal suffixes / punctuation,
then does bidirectional substring matching. Conservative: prefers no match
over a wrong match.
"""
from __future__ import annotations
import re
from functools import lru_cache
from src.config import load_companies

_SUFFIXES = re.compile(
    r"\b(inc|incorporated|ltd|limited|llc|llp|pvt|private|corp|corporation|"
    r"company|co|gmbh|sa|ag|plc|labs|technologies|tech|systems|solutions|"
    r"software|india|global)\b\.?",
    re.IGNORECASE,
)
_NON_ALNUM = re.compile(r"[^a-z0-9]+")


def _normalize(name: str) -> str:
    if not name:
        return ""
    s = name.lower()
    s = _SUFFIXES.sub("", s)
    s = _NON_ALNUM.sub(" ", s).strip()
    s = re.sub(r"\s+", " ", s)
    return s


@lru_cache(maxsize=1)
def _tracked_index() -> dict[str, str]:
    """Map normalized → canonical name, once per process."""
    return {_normalize(c.name): c.name for c in load_companies()}


def match_company(raw: str | None) -> str | None:
    """Return canonical company name if a tracked match is found, else None."""
    if not raw:
        return None
    norm = _normalize(raw)
    if not norm:
        return None
    idx = _tracked_index()

    # 1. exact normalized match
    if norm in idx:
        return idx[norm]

    # 2. bidirectional substring match (longer side wins ambiguity)
    candidates: list[tuple[int, str]] = []
    for tracked_norm, canonical in idx.items():
        if not tracked_norm:
            continue
        if tracked_norm in norm or norm in tracked_norm:
            # length-of-overlap-ish heuristic
            score = min(len(tracked_norm), len(norm))
            candidates.append((score, canonical))

    if not candidates:
        return None
    # take the highest-scoring candidate
    candidates.sort(reverse=True)
    return candidates[0][1]
